src/Symfony/Component/Security/Http/RememberMe/PersistentRememberMeHandler.php

```diff
@@ -0,0 +1,110 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Http\RememberMe;
+
+use Psr\Log\LoggerInterface;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\Security\Core\Authentication\RememberMe\PersistentToken;
+use Symfony\Component\Security\Core\Authentication\RememberMe\TokenProviderInterface;
+use Symfony\Component\Security\Core\Exception\AuthenticationException;
+use Symfony\Component\Security\Core\Exception\CookieTheftException;
+use Symfony\Component\Security\Core\User\UserInterface;
+use Symfony\Component\Security\Core\User\UserProviderInterface;
+
+/**
+ * Implements remember me tokens using a {@see TokenProviderInterface}.
+ *
+ * This requires storing remember me tokens in a database. This allows
+ * more control over the invalidation of remember me tokens. See
+ * {@see SignatureRememberMeHandler} if you don't want to use a database.
+ *
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ *
+ * @experimental in 5.3
+ */
+final class PersistentRememberMeHandler extends AbstractRememberMeHandler
+{
+    private $tokenProvider;
+    private $secret;
+
+    public function __construct(TokenProviderInterface $tokenProvider, string $secret, UserProviderInterface $userProvider, RequestStack $requestStack, array $options, ?LoggerInterface $logger = null)
+    {
+        parent::__construct($userProvider, $requestStack, $options, $logger);
+
+        $this->tokenProvider = $tokenProvider;
+        $this->secret = $secret;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createRememberMeCookie(UserInterface $user): void
+    {
+        $series = base64_encode(random_bytes(64));
+        $tokenValue = $this->generateHash(base64_encode(random_bytes(64)));
+        $token = new PersistentToken(\get_class($user), $user->getUsername(), $series, $tokenValue, new \DateTime());
+
+        $this->tokenProvider->createNewToken($token);
+        $this->createCookie(RememberMeDetails::fromPersistentToken($token, time() + $this->options['lifetime']));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function processRememberMe(RememberMeDetails $rememberMeDetails, UserInterface $user): void
+    {
+        [$series, $tokenValue] = explode(':', $rememberMeDetails->getValue());
+        $persistentToken = $this->tokenProvider->loadTokenBySeries($series);
+        if (!hash_equals($persistentToken->getTokenValue(), $tokenValue)) {
+            throw new CookieTheftException('This token was already used. The account is possibly compromised.');
+        }
+
+        if ($persistentToken->getLastUsed()->getTimestamp() + $this->options['lifetime'] < time()) {
```

Comment:
Does this mean that if I just continually "used" my cookie that I could get it to last longer than its lifetime? What is the meaning of lifetime? Like, if I set it to 1 week.. should it absolutely not be usable in 1 week? Or something different?

Reply:
A new remember me cookie is created each time the remember me is called (i.e. the session has expired). The ttl defines the time to the life from the last usage of the remember me. So e.g. 1 week means you're fully logged out after 1 week of inactivity, but you can continually use it if you use the website on a daily basis.

This is btw an exact copy of the current `PersistentTokenBasedRememberMeServices` logic.

Reply:
Ok, so this matches past functionality. So that's ok.

But I also think, perhaps not in this PR (or perhaps yes in this PR?) we should change that. When I set my remember me lifetime to 1 week, I would expect that it is valid for 1 week MAX - that a user (especially a bad user & especially when using non-persistent cookies) couldn't continue to re-use that indefinitely to access my account. It seems to me that the correct behavior is:

A) With persistent storage, we should store a `createdAt` date and use THAT when comparing the lifetime, not a `lastUsed` date.
B) For non-persistent storage (and i guess this probably also applies to persistent) the cookie should contain the lifetime / expiration in it. And this expiration should not *extend* when I use the cookie (even if we need to create a new remember me cookie when the session is expired, we should do so with a *shorter* expiration date, so that the new cookie expires at the same date as the original).

This is an important, and potentially controversial (? ) point. So I would be ok with merging the PR (as it matches the existing functionality) and opening an issue / PR to address this after.

Reply:
Hmm, let's indeed discuss this in an issue.

Looking at e.g. GitLab, they implement Remember me like we do currently. So I'm not sure which of the 2 is expected behavior.